implementations/rust/ockam/ockam_command/src/upgrade.rs

```diff
@@ -1,51 +1,149 @@
-struct UpgradeFile {
-    upgrade_message: Option<String>,
-    upgrade_message_macos: Option<String>,
-}
-
-pub fn check_if_an_upgrade_is_available() {
-    if !upgrade_check_is_disabled() {
-        // check if a new version has been released
-        Builder::new_current_thread()
-            .enable_all()
-            .build()
-            .unwrap()
-            .block_on(check());
+
+use crate::{
+    fmt_info, fmt_ok, node::util::spawn_node, terminal::ConfirmResult, CommandGlobalOpts, Result,
+};
+
+pub fn check_if_an_upgrade_is_available(opts: &CommandGlobalOpts) {
+    if upgrade_check_is_enabled() {
+        check_upgrade_sync(opts); // check if a new version has been released
     }
 }
 
-async fn check() {
-    let url = format!(
-        "https://github.com/build-trust/ockam/releases/download/ockam_v{}/upgrade.json",
-        crate_version!()
-    );
-    let resp = reqwest::get(url).await;
+fn upgrade_check_is_enabled() -> bool {
+    !get_env_with_default("OCKAM_DISABLE_UPGRADE_CHECK", false).unwrap_or(false)
+}
 
-    if let Ok(r) = resp {
-        if let Ok(upgrade) = r.json::<UpgradeFile>().await {
-            if let Some(message) = upgrade.upgrade_message {
-                eprintln!("\n{}", message.yellow());
+#[derive(Clone, Debug, Args)]
+#[command(
+    arg_required_else_help = false,
+    subcommand_required = false,
+    long_about = "Upgrade ockam to the latest version"
+)]
+pub struct UpgradeCommand {
+    #[arg(long, short)]
+    check: bool,
+    #[arg(long, short)]
+    yes: bool,
+}
 
-                if cfg!(target_os = "macos") {
-                    if let Some(message) = upgrade.upgrade_message_macos {
-                        eprintln!("\n{}", message.yellow());
-                    }
-                }
+impl UpgradeCommand {
+    pub fn run(self, opts: CommandGlobalOpts) {
+        local_cmd(run_impl(opts, self));
+    }
+}
 
-                eprintln!();
+fn run_impl(opts: CommandGlobalOpts, cmd: UpgradeCommand) -> miette::Result<()> {
+    let latest_release = get_latest_release_version_sync()?;
+    let latest_version = latest_release.version()?;
+    let current_version = crate_version!();
+
+    if current_version == latest_version {
+        opts.terminal
+            .stdout()
+            .plain(fmt_info!(
+                "You are already running the latest version of ockam: {}",
+                current_version
+            ))
+            .write_line()?;
+        return Ok(());
+    }
+
+    opts.terminal.write_line(fmt_info!(
+        "A new version of ockam is available: {}. Current ockam version: {}",
+        latest_version,
+        current_version
+    ))?;
+
+    if cmd.check {
+        return Ok(());
+    }
+
+    if !cmd.yes {
+        match opts.terminal.confirm(&fmt_info!(
+            "This will upgrade ockam to the latest version. Are you sure?"
+        ))? {
+            ConfirmResult::Yes => {}
+            ConfirmResult::No => {
+                return Ok(());
```

Comment:
What happens if the user doesn't want to upgrade? Will they see the warning every time they run a command?

Reply:
This will only run whenever they run `ockam upgrade`. Currently there is a check that is run on every command that looks that will warn you if you're out of date.

Reply:
Ok got it, thanks! 🙏